### ISSUE TITLE
[Bug] Unusual CPU overhead of SDPA call on H100

### ISSUE THREAD

**Issue by Jackmin801** [pytorch/torchtune#1652]
There seems to be something unusual about the performance characteristics of the attention implementation when running Llama models on H100. Namely that there is an unusually long overhead in the first sdpa forward and backward. This isnt observed in other cards (tested 4090 and 3090) which use the flash attention implementation.

![image](https://github.com/user-attachments/assets/1bc263f7-4ad9-4653-9579-bc90d84a3a5a)
![image](https://github.com/user-attachments/assets/b19aafe3-b8fe-4a08-a59f-15885f7c80f0)

A warning appears in the output about mistmatched strides in the output and grad output from the cuDNN SDPA bwd:
```bash
/root/miniconda3/lib/python3.12/site-packages/torch/autograd/graph.py:825: UserWarning: cuDNN SDPA backward got grad_output.strides() != output.strides(), attempting to materialize a grad_output with matching strides... (Triggered internally at ../aten/src/ATen/native/cudnn/MHA.cpp:674.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
```

#### 1. Clone repo at this branch
```bash
git clone --branch feat-activation-offloading-streams-dist git@github.com:Jackmin801/torchtune
cd torchtune
```
#### 2. Dev Install with torch nightly
```bash
pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu121
pip install torchao
pip install matplotlib pandas numpy
pip install -e ".[dev]"
```
#### 3. Login to huggingface and download model
```bash
huggingface-cli login --token <HF_TOKEN>
```
```bash
tune download meta-llama/Meta-Llama-3.1-8B-Instruct --output-dir /tmp/Meta-Llama-3.1-8B-Instruct --ignore-patterns "original/*.pth"
```
#### 4. Run single device
```bash
#tune run --nnodes 1 --nproc-per-node 2 lora_finetune_distributed --config recipes/configs/llama3_1/8B_lora.yaml \
tune run lora_finetune_single_device --config recipes/configs/llama3_1/8B_lora.yaml \
  optimizer_in_bwd=False \
  enable_activation_offloading=False \
  enable_activation_checkpointing=True \
  compile=False \
  model.lora_attn_modules="[q_proj,v_proj]" \
  model.apply_lora_to_mlp=False \
  model.apply_lora_to_output=False \
  model.lora_rank=8 \
  model.lora_alpha=16 \
  dataset.source=Yukang/LongAlpaca-12k \
  dataset.packed=False \
  dataset.split=train[:10%] \
  dataset.train_on_input=True \
  tokenizer.max_seq_len=8192 \
  metric_logger=torchtune.training.metric_logging.StdoutLogger \
  metric_logger.project=recipe_profiling \
  log_every_n_steps=1 \
  log_peak_memory_stats=True \
  gradient_accumulation_steps=1 \
  max_steps_per_epoch=4 \
  epochs=1 \
  batch_size=2 \
  metric_logger.name=llama3__qlora__seqlen_8192__act_ckpt_True__act_off_True__bs2 \
  profiler.enabled=True \
  profiler.profile_memory=True \
  profiler.with_stack=True \
  profiler.wait_steps=0 \
  profiler.warmup_steps=2 \
  profiler.active_steps=2 \
  profiler.num_cycles=1
```
#### 5. Observe the trace
Weird right?

**Comment by felipemello1** [comment]
thanks for reporting it! Quick question: Do you know if that happens in the very first batch vs in the first layer of every single batch?

Because if it happens only in the very first batch, then it may just be the time it takes to compile.

**Comment by Jackmin801** [comment]
It seems to occur every batch. hrmm dont think its about pos_embeds, otherwise it would happen for flash attn too?

**Comment by Jackmin801** [comment]
Im wondering if its a regression from cudnn so im building pytorch with older cudnn versions to see if anything changes

**Comment by Jackmin801** [comment]
Seems to be from the CUDNN_ATTENTION implementation of sdpa. With the `viable/strict` build of pytorch, you can toggle the bug by forcing FLASH or CUDNN implementation.

```python
    from torch.nn.attention import sdpa_kernel, SDPBackend
    with sdpa_kernel([SDPBackend.FLASH_ATTENTION]):
        recipe.train()
```
<img width="694" alt="image" src="https://github.com/user-attachments/assets/b6c10092-de9d-4d81-8b34-8a7d3ccb72ff">


```python
    from torch.nn.attention import sdpa_kernel, SDPBackend
    with sdpa_kernel([SDPBackend.CUDNN_ATTENTION]):
        recipe.train()
```

<img width="832" alt="image" src="https://github.com/user-attachments/assets/a1634980-bf2b-4de0-bad7-3027d5408699">

@eqy @drisspg @Skylion007

**Comment by eqy** [comment]
This overhead might be expected as the cuDNN SDPA backend is doing JIT compilation. Could you share a minimal example e.g., exercising the shape(s) that show the issue? We can work with cuDNN to try and get the overhead reduced.

**Comment by janeyx99** [comment]
Ah, it does only occur on the first call in the fwd and the first call in the bwd, so that caching would make sense.

**Comment by Jackmin801** [comment]
Ah that might be it. The shapes are different between the first batch and the second batch. So padding to max length should fix this right?

<img width="577" alt="image" src="https://github.com/user-attachments/assets/99a32251-ddd0-4aaa-83c9-4fcc5dcc3604">

**Comment by drisspg** [comment]
@eqy does CuDNN jit compile for every updated sequence length? That seems non ideal

**Comment by eqy** [comment]
@drisspg at present it would trigger recompilation but I'll reach to cuDNN to see if we can get reuse here
